File: packages/py/src/airlock/interface/runner.py
"""The agent runner: a driving adapter that resumes runs from the bus."""

from airlock.application.agent import Agent
from airlock.application.ports.event_bus import EventSubscriber
from airlock.domain.events import EventTopic
from airlock.interface.event_schemas import parse_approval_decided


class AgentRunner:
    """Subscribes to `approval.decided`, validates the payload, and calls
    `Agent.resume`. Holds no run state of its own (it lives in the store), so it
    can scale independently of the approvers."""

    def __init__(self, agent: Agent, subscriber: EventSubscriber) -> None:
        self._agent = agent
        self._subscriber = subscriber

    async def start(self) -> None:
        await self._subscriber.subscribe(EventTopic.APPROVAL_DECIDED, self._on_decision)

    async def _on_decision(self, event: object) -> None:
        decided = parse_approval_decided(event)
        await self._agent.resume(decided.run_id, decided.request_id, decided.decision)

File: packages/ts/src/interface/runner.ts
import type { Agent } from "../application/agent";
import type { EventSubscriber } from "../application/ports/event-bus";
import { EventTopic } from "../domain/events";
import { parseApprovalDecided } from "./event-schemas";

export interface AgentRunnerDependencies {
  readonly agent: Agent;
  readonly subscriber: EventSubscriber;
}

/**
 * Driving adapter that resumes runs. It subscribes to `approval.decided`,
 * validates the payload, and calls `Agent.resume`. The runner holds no run
 * state of its own (it lives in the store), so it is stateless between events
 * and can scale independently of the approvers.
 */
export class AgentRunner {
  private readonly agent: Agent;
  private readonly subscriber: EventSubscriber;

  constructor(deps: AgentRunnerDependencies) {
    this.agent = deps.agent;
    this.subscriber = deps.subscriber;
  }

  async start(): Promise<void> {
    await this.subscriber.subscribe(EventTopic.ApprovalDecided, (event) => this.onDecision(event));
  }

  private async onDecision(event: unknown): Promise<void> {
    const decided = parseApprovalDecided(event);
    await this.agent.resume(decided.runId, decided.requestId, decided.decision);
  }
}
